src/Common/OvercommitTracker.cpp

```diff
@@ -115,7 +115,14 @@ void OvercommitTracker::tryContinueQueryExecutionAfterFree(Int64 amount)
     if (OvercommitTrackerBlockerInThread::isBlocked())
         return;
 
-    std::lock_guard guard(overcommit_m);
+    {
+        std::shared_lock read_lock(overcommit_m);
+        if (cancellation_state == QueryCancellationState::NONE)
+            return;
+        read_lock.unlock();
+    }
+
+    std::lock_guard lk(overcommit_m);
```

Comment:
I would either remove the extra block or the explicit call to `unlock`. Otherwise it is confusing. Usually when we do something like this, we do it with the intention of manually unlock a lock to make sure the lock is unlocked before a specific object is destroyed. Here this is not the case.

I think you kind of mixed the two different approaches I suggested. I think you should choose one of them:
```
    {
        std::shared_lock read_lock(overcommit_m);
        if (cancellation_state == QueryCancellationState::NONE)
            return;
    }

    std::lock_guard lk(overcommit_m);
```

or

```
    std::shared_lock read_lock(overcommit_m);
    if (cancellation_state == QueryCancellationState::NONE)
        return;
    read_lock.unlock();

    std::lock_guard lk(overcommit_m);
````

The same for the other function.

Reply:
Thanks. I have remove the unlock in the extra block to keep the code simple.